Mathlib/Dynamics/SymbolicDynamics/Basic.lean

```diff
@@ -0,0 +1,602 @@
+definitions these set-theoretic identities are tautological.
+Thus the file develops the theory ambiently, and subshifts reuse it by restriction.
+
+**Working inside a subshift.**
+For `Y : Subshift A G`, cylinders and occurrence sets *inside `Y`* are simply
+preimages of the ambient ones under the inclusion `Y → (G → A)`. For example:
+
+`{ y : Y | ∀ i ∈ U, (y : G → A) i = (x : G → A) i } = (Subtype.val) ⁻¹’ (cylinder U (x : G → A)).`
+
+Shift invariance guarantees that the ambient shift restricts to `Y`.
+
+**Ergonomics.**
+Thin wrappers (e.g. `Subshift.shift`, `Subshift.cylinder`, `Subshift.languageOn`)
+may be added for convenience. They introduce no new theory and unfold to the
+ambient definitions.
+
+## Namespacing policy
+
+All ambient definitions live under the namespace `SymbolicDynamics.Ambiant`.
+If inner, subshift-relative wrappers are provided, they will be placed in the
+subnamespace `SymbolicDynamics.Inner`. This separation avoids name clashes
+between the two viewpoints, since both may naturally want to reuse names like
+`cylinder`, `shift`, `occursAt`, or `languageOn`.
+
+## Implementation notes
+
+* Openness/closedness results for cylinders and occurrence sets use
+  `[DiscreteTopology A]`. Closedness proofs that enumerate values additionally
+  require `[Fintype A]`, `[DecidableEq A]`, and `[DecidableEq G]` (for `Finset`
+  manipulations and `Function.update`).
+-/
+
+
+noncomputable section
+open Set Topology
+
+namespace SymbolicDynamics
+
+namespace Ambiant
+
+/-! ## Full shift and shift action -/
+
+section ShiftDefinition
+
+variable {A : Type*}
+variable {G : Type*}
+variable [Monoid G]
+
+/-- Right-translation shift:
+* In multiplicative notation: `(mulShift g x) h = x (h * g)`.
+* In additive notation (e.g. for `ℤ^d`): `(addShift v x) u = x (u + v)` -/
+@[to_additive]
+def mulShift (g : G) (x : G → A) : G → A :=
+  fun h => x (h * g)
+
+attribute [inherit_doc SymbolicDynamics.Ambiant.mulShift] SymbolicDynamics.Ambiant.addShift
+
+end ShiftDefinition
+
+section ShiftAlgebra
+
+variable {A G : Type*} [Monoid G]
+
+@[to_additive] lemma mulShift_apply (g h : G) (x : G → A) :
+  mulShift g x h = x (h * g) := rfl
+
+@[to_additive] lemma mulShift_one (x : G → A) : mulShift (1 : G) x = x := by
+  ext h; simp [mulShift]
+
+@[to_additive addShift_mul] lemma mulShift_mul (g₁ g₂ : G) (x : G → A) :
+  mulShift (g₁ * g₂) x = mulShift g₁ (mulShift g₂ x) := by
+  ext h; simp [mulShift, mul_assoc]
+
+end ShiftAlgebra
+
+section ShiftTopology  -- add only topology on A
+
+variable {A G : Type*} [Monoid G] [TopologicalSpace A]
+
+@[to_additive] lemma continuous_mulShift (g : G) :
+    Continuous (mulShift (A := A) (G := G) g) := by
+  -- coordinate projections are continuous; composition preserves continuity
+  continuity
+
+end ShiftTopology
+
+/-! ## Cylinders -/
+
+section CylindersDefs
+
+variable {A G : Type*}
+
+/-- Cylinder fixing `x` on the finite set `U`. -/
```

Comment:
This docstring is not really good, as people who don't know what a cylinder is can not guess it from the docstring. Could you try to improve it?